### ISSUE TITLE
FASTBuild scripting language design

### ISSUE THREAD

**Issue by clevijoki** [fastbuild/fastbuild#316]
# Overview

Hold onto your butts because this is going to be a long one. I find the build language to be quite awkward to express what I want to do.

I think I get the point of how it is intended to work, it tries to prevent cleverness and be stateless enough that you can statically find errors. It also tries to do this without becoming an indecipherable functional language. These are also goals I agree with. But I think the language is filled with avoidable gotchas and edge cases. It forces you to be clever to make things functional at all.

# No If Statement?

First off, you can hack in a fake 'if' statement with the 'ForEach' call:

```
.lib_wants_copy = {}

Using(.lib_properties)
ForEach(.check in .lib_optional_step) // equivalent to If(.lib_wants_copy)
{
    // do something like post-copy the library
}
```

You can even get extra clever with it, this can check if a list of values has something and then perform a step

```
.values_has_something = {}
ForEach(.value in .list_of_values)
{
    ^values_has_something = {'true'}
}

ForEach(.check in .values_has_something)
{
    // do stuff
}
```

That is just horrible and nobody should do this

## Suggestion

I would prefer to just have an actual if statement. I don't think the logic of if/else is really an issue to what I think are core goals. I don't really think we need a whole host of comparison operators, but being able to just check to see if a value is in a list would be very useful:

```
If(.platform in .supported_platforms)
{
    // do stuff
}
```

If/else is possible in a functional language like haskell, the real problem to avoid is state changes. Which brings me to...

# Using() creates implicit state dependent code

This statement:

```
ForEach(.platform in .PLATFORMS)
{
    Using(.platform)

    Print(.name)
}

```

You have no idea what .platform contains and just stomped, if .name was supposed to be from that struct or a local value. You also may run into accidental name collisions where multiple structs contain .name but that was also used by something else. This means you have to really start prefixing all variables with the 'type' name of the struct, like:

```
ForEach(.platform in .b_PLATFORMS) // some hint about that PLATFORMS might be in another struct
{
    Using(.platform)

    Print(.Platform_name) // this seems saner
    Print(.Config_name) // this becomes possible
}
```

If this also means you can get rid of the . syntax to find variables I also think that would help.

## Suggestion

What would really be better IMO is to just have a syntax to access a struct member:

```
ForEach(.platform in .b.PLATFORMS) // now you know where PLATFORMS comes from
{
    Print(.platform.name) // state stays immutable, tokens are shorter, where 'name' comes from is clearer
    Print(.b.name)
}
```

# ^ Operator gotchas

If I do this:

```
Libraries('a')
{
    ForEach(.include in .'a_$platform$_$config$_includes')
    {
        ^CompilerOptions + ' ' .'$platform$_include_flag' + .include
    }
}

Executable('b')
{
    // whoops just inherited a's compiler options
}
```

The "correct" fix is to perform this work-around every time you code things.

```
Executable('a')
{
    .extra_includes = {}
    ForEach(.include in .'a_$platform$_$config$_includes')
    {
        ^CompilerOptions + ' ' .'$platform$_include_flag' + .include
    }
    .CompilerOptions + .extra_includes
}
```

This is a gotcha I am sure many people ran into trying to use the language.

## Suggestion

Make ^ a global variable which lives at the point it was assigned

```
^some_var = 'values' // always works because it always assigns to 

Executable('a')
{
    ^extra_includes = {} // because it's an assignment there is a new value here
    ForEach(.include in .'a_$platform$_$config$_includes')
    {
        ^extra_includes + ' ' .'$platform$_include_flag' + .include
    }
    .CompilerOptions + ^extra_includes
}

Print(^extra_includes) // error variable extra_includes not found
```

# L-Value vs R-Value confusion

The next issue I have is just with how the language parser works internal is highly non-standard. With most languages someting like this should work:

```
ForEach(.config in {'Debug', 'Release', 'Final'})
{
    Print(.config)
}
```

It does not in FASTBuild. This should also work:

```
.ProjectConfigs = {[.Platform = 'Win32', .Config = 'Debug'], [.Platform = 'Win32', .Config = 'Release']}
```

This weirdness of differentiating between rvalues and lvalues extends to other aspects, like:

```
.target = 'lib-$platform$-$config$'
Library(.target) // error, expects Library('$target$')
{
    
}
```

## Suggestion

Just make it work like a normal language

# Executable vs DLL vs Linker

I am very interested in using FASTBuild to solve content compilation problems. e.g. I want to be able to compress textures from .tga to target platform files. I could likely fake the 'scanner' step by making my texture compressor support /showIncludes (if I read some additional compression settings file for example). I could also work around the lack of multiple output extensions being declared by forcing things to wait on aliases.

But what really feels awkward about this all is to perform a linker stage, e.g. perform a many -> one process, I need to somehow hack this into Executable, DLL, or Library, despite them all being extremely similar.

## Suggestion

Remove them all and just replace them with a single Linker() function. Platform specific hacks like whatever makes them different now can still be done via arguments. 

# Scoping Hell

I have issues inherent to trying to express things with scoping. Let's say I have a project structure like this:

```
Source/
    External/
        wxWidgets/
        SDL/
        jpeg/
        png/
        vorbis/
        External.bff

    Game/
        Client/
        Server/
        Game.bff

    Tools/
        Editor/
        CustomCompiler/
        Tools.bff
fbuild.bff
```

Now I want to build the game for Win32/Win64/OSX/Linux64, with the platforms Debug/Release/Profile/Final. I want to build the tools only for Win64 and Debug/Release. I also want to generate a visual studio solution

How do you do this by minimizing typing? This is one guess at what seems like an 'obvious' approach. This is also what I see recommended in other posts:

```
.ALL_TARGETS = {}
ForEach(.platform in .ALL_PLATFORMS)
{
    ForEach(.config in .ALL_CONFIGS)
    {
        .VARIANT = '$platform$-$config$' // build variant
        .OUTPUTDIR = 'build/$platform$/$config$'

        Using(.'$platform$_CppSetup_$config$')

        #include "Source/External/External.bff"
        #include "Source/Game/Game.bff"
    }
}

ForEach(.platform in .TOOLS_PLATFORMS)
{
    ForEach(.config in .TOOLS_CONFIGS)
    {
        // why is all of this done twice?

        .VARIANT = '$platform$-$config$'
        .OUTPUTDIR = 'build/$platform$/$config$'

        Using(.'$platform$_CppSetup_$config$')

        #include "Source/Tools/Tools.bff"
    }
}

Alias('all') { .Targets = .ALL_TARGETS }
```

This means that 'Game.bff' can be simple and use FASTBuild syntax directly like:

```
ObjectList('game-client-objlist-$VARIANT$')
{
    .CompilerInputPath = 'Source/Game/Client'
    .CompilerOutputPath = '$OUTPUTDIR$/$CompilerInputPath$'
}

Executable('game-client')
{
    .CompilerInputPath = 'Source/Game/Client'
    .CompilerOutputPath = '$OUTPUTDIR$/$CompilerOutputPath$/GameClient.$EXE_EXT$'
}
^ALL_TARGETS + 'game-client'
```

... but now there are some problems too.. I've now defined wxWidgets to build under too many platforms. Do I need to have a ToolsExternal.bff? Also where do I define per-platform steps, like if I wanted to create a VS solution for Win64. They can't go into that loop because they would be defined per variant. I would need to put them into another .bff file like ToolsVCProjects.bff, but that means that the setup for the two logically connected things has to be separated.

I could do this:

```
// GAME_Executable.bff
ForEach(.platform in .GAME_PLATFORMS) 
{ 
    ForEach(.config in .GAME_CONFIGS) 
    {
        Using(.'$platform$_ExeCppSetup_$config$')
        .CompilerOutputPath = '$Build$/$platform$/$config$/$CompilerInputPath$/$target$.' + '.$platform$_exe_ext'
        .target = '$TargetName$-$platform$-$config$'
        ObjectList('objlist-$target$') {}
        Executable('$target$') { .Libraries + 'objlist-$target$' } 
        ^ALL_TARGETS + '$TargetName$-$platform$-$config$'
    }
}
// repeat for each type I want

// Now in my configuration scripts I can do

// vorbis
{
    .TargetName = 'vorbis'
    .CompilerInputPath = 'Source/External/vorbis/src'

    #include "GAME_Library.bff"
    #include "VCXProject.bff"
}

// png
{
    .TargetName = 'png'
    .CompilerInputPath = 'Source/External/png/src'

    #include "GAME_Library.bff"
    #include "VCXProject.bff"
}
```

What I have really done is produce a poor man's function call. It also doesn't solve the issue of I've just turned things into spaghetti.

## Suggestion

Add real function calls.

```
DefineFunction("CreateGameExecutable", .params) {

    ForEach(.platform in .GAME_PLATFORMS) 
    { 
        ForEach(.config in .GAME_CONFIGS) 
        {
            .use_params = .'$platform$_ExeCppSetup_$config$' + .params
            .target = '$TargetName$-$platform$-$config$'

            Using(.use_params) // without Using you could just assign these manually, or say ObjectList(target, .use_params) or something.

            ObjectList('objlist-$target$') 
            {
                .CompilerOutputPath = '$Build$/$platform$/$config$/$CompilerInputPath$/$target$
            }
            Executable('$target$') 
            { 
                .CompilerOutputPath = '$Build$/$platform$/$config$/$CompilerInputPath$/$target$.' + '.$platform$_exe_ext'
                .Libraries + 'objlist-$target$' 
            }

            ^ALL_TARGETS + '$TargetName$-$platform$-$config$'
        }
    }
}
```

You should be able to just call functions with a minor change to the syntax (with r-value support) to keep the current look/feel:

```
CreateGameExecutable([
        .Target = 'target_name'
        .CompilerOutputPath = 'path'
    ])
```

With all of the above strictness changes you may even be able to print an error automatically if you reference a value that wasn't found in the input params. Optional params could be implemented like this:

```
DefineFunction("MyFunc", .params)
{
    .use_params = [.optional = ''] + .params

    // now .use_params.optional will never be an error
}
```

# Summary

If these changes were to be done I would just want to have a new extension like .bf2 or something. I would also be interested in assisting with this if people thought the syntax changes were heading in a good direction.

I think the most important quality a language can have is ease of learning, and when things work just a bit weird and have gotchas that really hinders adoption.

**Comment by clevijoki** [comment]
Another example of needing an 'if' statement is how do I add compiler-specific options to one platform or config? Like say I want to disable a specific warning on that platform,  /wd4127. If I have:
```
ForEach(.platform in .Platforms)
{
    ForEach(.config in .Configs)
    {
        Using(.'$platform$_$config$_CppSetup')

        ObjectList('objectlist-somelib-$platform$-$config$')
        {
            // now I just want to add /wd4127 to this ObjectList only if under Win32 or Win64 platforms. How is this possible to do?
        }
    }
}
```

**Comment by dummyunit** [comment]
When If-statement is only needed to modify a variable 1 based on a value of variable 2 that can be achieved by creating a "map" of "value of variable 2" -> "desired addition to variable 1":
```
.CompilerOptions_Win32 = '/wd4127'
.CompilerOptions_Win64 = '/wd4127'
.CompilerOptions_Linux32 = ''
.CompilerOptions_Linux64 = ''
.CompilerOptions_SomeOtherPlatform = ''
ForEach(.platform in .Platforms)
{
    ForEach(.config in .Configs)
    {
        Using(.'$platform$_$config$_CppSetup')
        ObjectList('objectlist-somelib-$platform$-$config$')
        {
            .CompilerOptions + .'CompilerOptions_$platform$'
        }
    }
}
```
Only downside is that you have to create `.CompilerOption_` for all platforms even when you want a special handling only for one of them. But, of course, real If statement would be better.

For me a more compelling case for If-statement is when you need to call a different bff-function depending on variable value. I have one such case when we have to provide one library as a DLL instead on linking it in the EXE in one specific configuration. (So a need to call `DLL` instead of `StaticLibrary`).

**Comment by clevijoki** [comment]
That's kind of what I mean by the language forces you to have to be clever to solve problems. That will make adding a new platform tedious to update every package with platform specific warnings.

Another way to solve things would be to have some way to get an optional variable:

```
.CompilerOptions + ?'CompilerOptions_$platform$'
``` 
.. or something. It can evaluate to the expected empty value if not found.

Having an If statement with scoping rules would actually be worse in this case because you run into the ^ global modification problems without a local workaround.

**Comment by ffulin** [comment]
@clevijoki Thanks for your detailed thoughts.

The TL;DR: is that I agree with most (almost all) of what you've said, but I'll provide some details about each of your points below. Please let me know if I can better explain any of it.

**History of the BFF syntax**

The language was intended to meet a few of goals:
 - not need parsing if the config files don't change
 - being build-domain specific: optimized for expressing build structure succinctly (as opposed to being a general scripting language)
 - avoiding the ability to implement arbitrary (and in particular non-deterministic) logic
 - being easy to use with friendly, well documented errors
 - being fast to parse

For the most part I think the language is ok, but doesn't have all the features it needs to be great. It is capable of expressing very complex build structures, but there definitely are cases where it makes you jump through hoops in ways that can be annoying and/or hard to maintain. I have been overly cautious about adding new features and syntax to the core language to not hastily implement something I would later regret, but I definitely see the need for many improvements.

**If Statement**

The language definitely needs an if statement. It is something I would like to add, and I don't see any issues with supporting it.

There are several ways you can achieve the end result, with varying degrees of annoyance. I list them here for completeness, not to suggest that they negate the usefulness of a proper If:
 - constructing variable names (as mentioned by dummyunit)
 - #if directive (when the host platform and target are equal)
 - Your crazy thing which I agree nobody should do :)
 - parallel ForEach loops (you can create additional arrays of structures and iterate them in parallel using, as such: ForEach( .Config in .Configs, .ExtraArgs in .ExtraArgsPerConfig )
 - extending/overriding the configs locally and looping through those

Even for the ones that don't feel truly awful (like the last two), they are generally quite verbose and thus annoying for really simple things.

There are probably more ways to "abuse" the language to achieve the equivalent of an If, but a real If would just give the desired result.

**Using**

I actually find the Using syntax fairly logical, and when used in a disciplined way, quite maintainable. Having said that, I think the syntax could generally be extended to support scope in the way you suggested. That would give some additional flexibility and control, and allow you to be as precise as you wish.

**Operator ^**

I might be missing your point, but operator ^ does work in the way you describe in your example, the only difference being that you use the .extra_includes to declare the variable at the desired scope.

Not related to the operator ^ specifically, but this pattern (concatenating things in a loop, particularly strings) comes up reasonably often. An specific syntax to do that might be warranted. For example, imagine something along the lines of:

```C++
Exectuable( 'a' )
{
    .CompilerOptions + Expand(.includes, ' -I$1$')
```

This could be done as a built-in feature, or with functions/templates (see below).

**L-Value vs R-Value**

The internal implementation is indeed inconsistent in some places, resulting in syntax you would expect to work not working. I think it can continue to be improved and/or restructured internally to achieve the consistency you'd expect, supporting both of your examples (and others). I don't think there is anything fundamentally incompatible in the implementation - it just needs improving.

**Executable vs DLL vs Linker**

Internally, Executable and DLL are actually both the same thing (a reflection of LinkerNode), with the only difference being an "isDLL" flag that activates some behavioral changes and extra error checks.  Executable and DLL are provided as distinct options for convenience. The underlying Linker could directly be exposed if needed.

For your goal of running an arbitrary executable, I think you'd actually want to use Exec(), but that is missing some functionality (like good support for multiple inputs expanded on the command line) that would make it ideal for that. I'm actually considering adding those right now, as I have a use-case that could benefit from that.

**Scoping Hell**

I think your approach can work, but what I usually suggest instead, is defining a list of configs, which include the platform/config combinations you want (instead of two nested loops). Then I usually include each project (recursively or not depending on preference) and having each project loop through those configs internally. It's slightly more verbose, but is more flexible and allows each project to override any part of any config it needs.

**Function calls**

#including a bff in the examples you gave in the Scoping Hell section is indeed a (very) poor-man's function call. It's a crutch for the lack of a proper function call/templating system. This is something I would like to add to the language. I haven't locked down the exact syntax, but I am very much thinking along the lines of your example.

**.bf2 extension**

I actually think all the things you have mentioned can be fixed in, or added to, the existing syntax without the need for a new extension. I don't see anything that is not backwards compatible (with the exception of your ^ suggestion - though I think we don't want/need that, particularly if we address the other things).

**Finally**

> I think the most important quality a language can have is ease of learning, and when things work just a bit weird and have gotchas that really hinders adoption.

I totally agree with you here.  I want the language to be as logical and easy to use as possible. Fixing some of the issues you identified will help with that, as will improving some of the error reporting.

**Comment by g3cko** [comment]
One thing that I feel is a bit lacking (or I'm overlooking) - is manipulating file paths.  I keep thinking, maybe some style of path modifiers might make sense...  I know the above mentions "if statement" support, but I'm not sure that solves things or just allows me to work around it a little differently.  I do have work arounds, but just to share... 

Although I am generating bff files that allow me to customize quite a bit, I can't seem to set up my ForEach platform, ForEach config executable to easily allow what I want

```
ForEach(.Plat in .SupportedPlatforms)   
    ForEach(.Config in .SupportedConfigs)   
            .LinkerOutput = '$BinDir$/app.$ExeExtension$'   
```
Where - the platform defines the executable extension, and the output dir is also named based on combining the platform and config...  I want to easily change the %2  from app.exe to app.pdb.  I tried not adding the ".exe" to the LinkerOutput and adding it to the "%2.exe" but that doesn't work as it breaks the dependency tracking (tool was successful but no output found)


```
    .LinkerOptions      + ' /OUT:"%2"'
                        + ' /pdb:"%2.pdb"'
```
```
     // some sort of path modifying syntax to give me the 'root'  
    .LinkerOptions      + ' /OUT:"%2"'    
                        + ' /pdb:"expand(%2:r).pdb"'     

```
		Modifiers: (from vim)
			:p		expand to full path
			:h		head (last path component removed)
			:t		tail (last path component only)
			:r		root (one extension removed)
			:e		extension only

Another example is specific to when I use 'CompilerInputPathRecurse' for some code generation, I have a harder time getting the paths I want.  We have a tool that takes as input *.msg files and using the my root dir as the input path and recursing, I easily find them all and easily create an output path for them ("outputdir/partial/path.outputextension") .  However, the code gen tool also wanted the "output directory" - which I couldn't think of an easy way to get.  

This was easily worked around since it's a python based tool we wrote internally and I could simply make it support a output directory or output path, but I wanted to share this too

**Comment by poppolopoppo** [comment]
@g3cko totally agree for string modifiers.

I would also love a '/' operator to concatenate paths while staying platform agnostic :
````
.path = 'source' / 'output' / 'file.ext'
; Windows = 'source\output\file.ext'
; POSIX   = 'source/output/file.ext'
````
---
@clevijoki it just so happens that I solved this particular issue using the '?' optional assignment operator I proposed in #20 a while ago. 
Of course it's not available in vanilla fastbuild but I think it's still worth sharing the idiom :
````
// MyProjects.bff
//------------------------------------------------------------------------------

.Compile_Win32_Debug    = [ .Platform = 'Win32' .Config = 'Debug'   ]
.Compile_Win32_Release  = [ .Platform = 'Win32' .Config = 'Release' ]
.Compile_Win64_Debug    = [ .Platform = 'Win64' .Config = 'Debug'   ]
.Compile_Win64_Release  = [ .Platform = 'Win64' .Config = 'Release' ]

; Minimal compile configurations list
.CompileConfigs        = {
    .Compile_Win32_Debug,
    .Compile_Win32_Release,
    .Compile_Win64_Debug,
    .Compile_Win64_Release,
}

; Projects list
{
    .ProjectName = 'A'
    .ProjectLibraries_Win32 = { 'libfoo32' }
    .ProjectLibraries_Win64 = { 'libfoo64' }
    .ProjectLibraries_Win64_Debug = { 'libfoo64d' }
    #include "CompileProject.bff"
}
{
    .ProjectName = 'B'
    .ProjectLibraries = { 'libbar' }
    .ProjectLibraries_Debug = { 'libbard' }
    .ProjectLibrariesAdditional_Win32 = { 'libwin32' }
    #include "CompileProject.bff"
}
{
    .ProjectName = 'C'
    #include "CompileProject.bff"
}
````
````
// CompileProject.bff
// Insert your typical ObjectList helper
//------------------------------------------------------------------------------
{
    Print( '[Compile project $ProjectName$]' )

    ; Arguments default value
    .ProjectLibraries ? {}
    .ProjectLibrariesAdditional ? {} ; lets you specialize on 2 axes (Platform|Config)

    ; Insert specializable parameters here
    .EvalParameters = {
        'ProjectLibraries',
        'ProjectLibrariesAdditional'
        ; ...
    }

    ; Iterate over each compile config and do things
    ForEach( .CompileConfig in .CompileConfigs )
    {
        Using( .CompileConfig )

        #include "EvalParameters.bff"

        Print( '$ProjectName$_$Platform$_$Config$' )

        .ProjectLibraries + .ProjectLibrariesAdditional
        
        ForEach( .Library in .ProjectLibraries )
        {
            Print( ' + $Library$' )
        }
    }
}
````
````
// EvalParameters.bff
// Eval the list of parameters matching the platform and current config
//------------------------------------------------------------------------------

.EvalResults = []

ForEach( .Parameter in .EvalParameters )
{
    .Evaluation =
    [
        // The value chosen will be the most specialized one :
        ."$Parameter$_$Platform$"           ? ."$Parameter$"
        ."$Parameter$_$Config$"             ? ."$Parameter$_$Platform$"
        ."$Parameter$_$Platform$_$Config$"  ? ."$Parameter$_$Config$"
        ."$Parameter$"                      = ."$Parameter$_$Platform$_$Config$"
    ]
    ^EvalResults + .Evaluation
}

Using( .EvalResults )
````
When parsing this fastbuild will print (again using #20) :
````
[Compile project A]
A_Win32_Debug
 + libfoo32
A_Win32_Release
 + libfoo32
A_Win64_Debug
 + libfoo64d
A_Win64_Release
 + libfoo64
[Compile project B]
B_Win32_Debug
 + libbard
 + libwin32
B_Win32_Release
 + libbar
 + libwin32
B_Win64_Debug
 + libbard
B_Win64_Release
 + libbar
[Compile project C]
C_Win32_Debug
C_Win32_Release
C_Win64_Debug
C_Win64_Release
````
This let you write only what you truly want to specialize in your projects definition, making it compact and maintainable since now you don't have to enumerate every combination of [platform, config] pairs everywhere. It solved many -if not all- of my problems regarding heterogeneous project configuration.

**Comment by Sigvaard** [comment]
I think it would be neater if fastbuild used existing language like lua instead, it would be easier for people to contribute and it would allow doing much more than .bff

**Comment by ffulin** [comment]
One thing for sure that FASTBuild will never have is a generic scripting language front-end like lua. The very flexibility that something like lua provides means that we'd have to run the configuration script every time you build because it could generate config based on pretty much anything and FASTBuild would have no sane way to know if it needed running or not. This would put us in the same place as most other build systems - where builds that build nothing still take a long time just to parse the config.

The plan is still to continue to extend the language along the lines of many/most of the things discussed in this thread.

**Comment by bugproof** [comment]
> The very flexibility that something like lua provides means that we'd have to run the configuration script every time you build because it could generate config based on pretty much anything and FASTBuild would have no sane way to know if it needed running or not.

You can still make it run only when it changed or whatever you do in fastbuild now, Lua is pretty extensible and you can write it in more declarative fashion (see premake for example). The flexibility is the key here. 

I was using fastbuild before and it was amazing and I would like to use it again but maintainability of BFF files was very hard especially after a few months when you needed to change something and it's very limiting so if you wanted more maintainable solution then you'd have to use premake/CMake and generate BFF files. 

I was even thinking to maintain a fork of fastbuild with Lua instead but currently lacking time to try doing that.

**Comment by missmah** [comment]
> One thing that I feel is a bit lacking (or I'm overlooking) - is manipulating file paths. I keep thinking, maybe some style of path modifiers might make sense... I know the above mentions "if statement" support, but I'm not sure that solves things or just allows me to work around it a little differently. I do have work arounds, but just to share...
> 
> Although I am generating bff files that allow me to customize quite a bit, I can't seem to set up my ForEach platform, ForEach config executable to easily allow what I want
> 
> ```
> ForEach(.Plat in .SupportedPlatforms)   
>     ForEach(.Config in .SupportedConfigs)   
>             .LinkerOutput = '$BinDir$/app.$ExeExtension$'   
> ```
> 
> Where - the platform defines the executable extension, and the output dir is also named based on combining the platform and config... I want to easily change the %2 from app.exe to app.pdb. I tried not adding the ".exe" to the LinkerOutput and adding it to the "%2.exe" but that doesn't work as it breaks the dependency tracking (tool was successful but no output found)
> 
> ```
>     .LinkerOptions      + ' /OUT:"%2"'
>                         + ' /pdb:"%2.pdb"'
> ```
> 
> ```
>      // some sort of path modifying syntax to give me the 'root'  
>     .LinkerOptions      + ' /OUT:"%2"'    
>                         + ' /pdb:"expand(%2:r).pdb"'     
> ```
> 
> ```
> 	Modifiers: (from vim)
> 		:p		expand to full path
> 		:h		head (last path component removed)
> 		:t		tail (last path component only)
> 		:r		root (one extension removed)
> 		:e		extension only
> ```
> 
> Another example is specific to when I use 'CompilerInputPathRecurse' for some code generation, I have a harder time getting the paths I want. We have a tool that takes as input *.msg files and using the my root dir as the input path and recursing, I easily find them all and easily create an output path for them ("outputdir/partial/path.outputextension") . However, the code gen tool also wanted the "output directory" - which I couldn't think of an easy way to get.
> 
> This was easily worked around since it's a python based tool we wrote internally and I could simply make it support a output directory or output path, but I wanted to share this too

I'm looking for exactly this. @ffulin I'd like to take `ObjectList`'s `%1` (or `%2`) and get the `filename` without the `path` attached. I don't think there's any way to do this currently in `fastbuild`, which is kind of a bit surprising/disappointing. Do you have any suggestions?

The closest I can think of is if I *know* the root path, maybe `subtraction` will work? But this won't work for files which are nested in sub-directories and merely found by the `ObjectList` pattern matching, because I won't know the `full path` precisely (only the `root` of the `path`).

**Comment by ffulin** [comment]
This issue is covers a lot of topics with some very useful discussions, but because it's so wide in coverage, is next to impossible to resolve.

Several items discusses here have since been resolved or implemented, or are covered by other more specific tickets.

In the interests of trying to get the backlog under control, I'm going to close this issue.

We can create new tickets for individual features or improvements to existing features as needed going forward, for anything that's not already covered.
